src/vs/editor/browser/viewParts/blockDecorations/blockDecorations.ts

```diff
@@ -0,0 +1,98 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { createFastDomNode, FastDomNode } from 'vs/base/browser/fastDomNode';
+import 'vs/css!./blockDecorations';
+import { RenderingContext, RestrictedRenderingContext } from 'vs/editor/browser/view/renderingContext';
+import { ViewPart } from 'vs/editor/browser/view/viewPart';
+import { EditorOption } from 'vs/editor/common/config/editorOptions';
+import * as viewEvents from 'vs/editor/common/viewEvents';
+import { ViewContext } from 'vs/editor/common/viewModel/viewContext';
+
+export class BlockDecorations extends ViewPart {
+
+	public domNode: FastDomNode<HTMLElement>;
+
+	private readonly blocks: FastDomNode<HTMLElement>[] = [];
+
+	private contentWidth: number = -1;
+
+	constructor(context: ViewContext) {
+		super(context);
+
+		this.domNode = createFastDomNode<HTMLElement>(document.createElement('div'));
+		this.domNode.setAttribute('role', 'presentation');
+		this.domNode.setAttribute('aria-hidden', 'true');
+		this.domNode.setClassName('blockDecorations-container');
+
+		this.update();
+	}
+
+	private update(): boolean {
+		let didChange = false;
+		const options = this._context.configuration.options;
+		const layoutInfo = options.get(EditorOption.layoutInfo);
+		const newContentWidth = layoutInfo.contentWidth - layoutInfo.verticalScrollbarWidth;
+
+		if (this.contentWidth !== newContentWidth) {
+			this.contentWidth = newContentWidth;
+			didChange = true;
+		}
+
+		return didChange;
+	}
+
+	public override dispose(): void {
+		super.dispose();
+	}
+
+	// --- begin event handlers
+
+	public override onConfigurationChanged(e: viewEvents.ViewConfigurationChangedEvent): boolean {
+		return this.update();
+	}
+	public override onScrollChanged(e: viewEvents.ViewScrollChangedEvent): boolean {
+		return e.scrollTopChanged || e.scrollHeightChanged || e.scrollLeftChanged;
+	}
+	public override onDecorationsChanged(e: viewEvents.ViewDecorationsChangedEvent): boolean {
+		return true;
+	}
+
+	// --- end event handlers
+	public prepareRender(ctx: RenderingContext): void {
+		// Nothing to read
+	}
+
+	public render(ctx: RestrictedRenderingContext): void {
+		let count = 0;
+		const decorations = ctx.getDecorationsInViewport();
+		for (const decoration of decorations) {
+			if (!decoration.options.blockClassName) {
+				continue;
+			}
+
+			let block = this.blocks[count];
+			if (!block) {
+				block = this.blocks[count] = createFastDomNode(document.createElement('div'));
+				this.domNode.appendChild(block);
+			}
+			const top = ctx.getVerticalOffsetForLineNumber(decoration.range.startLineNumber);
+			const bottom = ctx.getVerticalOffsetForLineNumber(decoration.range.endLineNumber + 1);
```

Comment:
This will contain any view zone immediately following this line. If you would like to limit the block strictly to the end line number, you would need to use something like `ctx.getVerticalOffsetForLineNumber(decoration.range.endLineNumber) + lineHeight`. `lineHeight` can be found in the editor options.

Reply:
Right, good catch. Unfortunately, this is the behavior I need, to cover deleted ranges in the 3wm editor.

I suggest to add a comment and merge it as this and solve the problem when someone does not want to include view zones.